gamemodes/terrortown/gamemode/shared/sh_equip_items.lua

```diff
@@ -876,43 +870,99 @@ local function ValueToKey(tbl)
 end
 
 ---
--- Cleans up the structure of the CanBuy table of all weapons and items.
+-- Cleans up the structure of the CanBuy table of the given equipment.
 -- After calling this, all keys will be equal to their value. This allows access in O(1) rather than O(n).
 -- table.HasValue is still supported, but please just check if the key is not nil instead.
+-- @param table eq Equipment you clean the buy indices from
 -- @internal
 -- @realm shared
-local function CleanUpDefaultCanBuyIndices()
-	local itms = items.GetList()
+-- @local
+local function CleanUpDefaultCanBuyIndices(eq)
+	eq.CanBuy = eq.CanBuy or {}
 
-	-- load items
-	for i = 1, #itms do
-		local itm = itms[i]
+	ValueToKey(eq.CanBuy)
+end
 
-		itm.CanBuy = itm.CanBuy or {}
+---
+-- Initializes the default equipment for traitors and detectives
+-- @param table eq Equipment you initialize
+-- @internal
+-- @realm shared
+function InitDefaultEquipment(eq)
+	CleanUpDefaultCanBuyIndices(eq)
+	InitDefaultEquipmentForRole(TRAITOR, eq)
+	InitDefaultEquipmentForRole(DETECTIVE, eq)
+end
 
-		ValueToKey(itm.CanBuy)
-	end
+---
+-- Resets the default equipment for a role
+-- @param roleData roleData Role you reset for
+-- @param table eq Equipment you reset
+-- @internal
+-- @realm shared
+local function ResetDefaultEquipmentForRole(roleData, eq)
+	local tbl = roleData.fallbackTable or {}
+	local tblSize = #tbl
+
+	-- is a buyable equipment to load info from
+	if not eq or eq.Doublicated then return end
+
+	local base = GetEquipmentBase(eq)
+	if not base then return end
 
-	local sweps = weapons.GetList()
+	local addEquipment = tobool(eq.CanBuy and eq.CanBuy[roleData.index])
 
-	-- load sweps
-	for i = 1, #sweps do
-		local wep = sweps[i]
+	if addEquipment then
+		local tableHasValue = false
 
-		wep.CanBuy = wep.CanBuy or {}
+		for i = 1, tblSize do
+			if tbl[i].id == base.id then
+				tableHasValue = true
 
-		ValueToKey(wep.CanBuy)
+				break
+			end
+		end
+
+		if not tableHasValue then
+			tbl[tblSize + 1] = base
+		end
+	else
+		local counter = 0
+
+		for i = 1, tblSize do
+			if tbl[i].id == base.id then continue end
+
+			counter = counter + 1
+			tbl[counter] = tbl[i]
+		end
+
+		local diff = tblSize - counter
+
+		if diff > 0 then
+			for i = 1, diff do
+				table.remove(tbl)
+			end
+		end
```

Comment:
Why are you doing that / what should this do?
Should this remove the equipment from the table if it exists? 
If so then this should not be based on counters / the amount and instead remove the item once the match is found. (Also this seems to be wrong atm)

Reply:
At the moment I wrote it, I wasnt aure how the table exactly was structured and therefore went the safe way and iterated over all items to remove them. But I guess this is unnecessary if you can rely on the fact that an item is at max one time inside the table. This generally cleans the ShopFallback up if after a hotreload the default loadout changes. Maybe you could do it more efficient by comparing shop-default changes before hotreload.
But this is generally more complicated than it needs to be, as I wasnt sure first why hotreloads dont work. Probably will rework this. Otherwise it only happens if you change a weapon live, so speed is not of the highest importance. This is safe but not efficient.

Reply:
I think thats fine with iterating over the table, but i do not get why you are sure that all instances of the equipment are at the end of the table and that you can just work with a counter. To me this can just be removed while looping through all items instead of continuing.

Reply:
Ahh okay, yeah I am still unsure about how tables are working inside lua and as it is an indexed table I was afraid, that when I remove an item, that there are gonna be missing indexes. So I wrote my own remover, that just skips all duplicated items and because the counter only increases when it is not continued, those places get filled with non-skipped items.
tbl[counter] = tbl[i]
so I only iterate once over it (like one iteration of bubble sort) and at the end, I know, that the rest is free for clearing. Removing at the end can be done in O(1) instead of O(n). And removing while iterating would lead to changing table size, which would also need to be handled with something else thats not a for-loop.

In general you could just remove those you find, but then the gmod table.remove function would do the same I did, but for the whole table for every item you find.

I thought this would the most efficient way, if you would need to remove more than one index.